### DISCUSSION TITLE
What's the best way to add to the SDL *before* codegen runs

### DISCUSSION THREAD

**Discussion by probablykabari** [dotansimha/graphql-code-generator#8505] [Q&A] [answered]
I have code that dynamically generates a few types based on a directive in the schema. We use an SDL-first approach so we need to read the schema and print this out before the rest of codegen.

Is this best done in one of the lifecycle hooks? Or can I do this in a plugin? (Plugin is preferred but I don't see how to add to the schema that's already been loaded.)

**Comment by dotansimha** [comment, marked as answer]
Codegen `plugin`s are only meant to deal with the output, and not with the input (schema/operations). 
Do you have an example of what schema change you need to perform? I guess ideally you should run your script and create a new SDL file with the modified content, and then run codegen on this file?

**Reply by probablykabari** [reply]
My specific use case is we wrote a directive that generates an InputType used in queries on the server side. The conundrum is—and why we didn't implement this using `mapSchema` or `wrapSchema`—is that if we're using SDL first but then have types that don't exist until the server is started we can satisfy codegen by pointing at the server but it doesn't satisfy the linter looking at the SDL files.

For now I opted to generate a new SDL before codegen runs but am looking into making it more automated.
